networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py

```diff
@@ -0,0 +1,241 @@
+import json
+import os
+
+import eventlet
+import netaddr
+from networking_nsxv3.common.constants import *
+from networking_nsxv3.common.locking import LockManager
+from networking_nsxv3.plugins.ml2.drivers.nsxv3.agent import provider_nsx_mgmt
+from networking_nsxv3.plugins.ml2.drivers.nsxv3.agent.constants_nsx import *
+from oslo_config import cfg
+from oslo_log import log as logging
+
+LOG = logging.getLogger(__name__)
+
+
+class API(provider_nsx_mgmt.API):
+    POLICY_BASE = "/policy/api/v1"
+
+    POLICIES = "/policy/api/v1/infra/domains/default/security-policies"
+    POLICY = "/policy/api/v1/infra/domains/default/security-policies/{}"
+
+    GROUPS = "/policy/api/v1/infra/domains/default/groups"
+    GROUP = "/policy/api/v1/infra/domains/default/groups/{}"
+
+    STATUS = "/policy/api/v1/infra/realized-state/status"
+
+
+class Payload(provider_nsx_mgmt.Payload):
+
+    def sg_members_container(self, os_sg, provider_sg):
+        sg = {
+            "display_name": os_sg.get("id"),
+            "expression": [
+                {
+                    "value": "security_group|{}".format(os_sg.get("id")),
+                    "member_type": "LogicalPort",
+                    "key": "Tag",
+                    "operator": "EQUALS",
+                    "resource_type": "Condition"
+                }
+            ],
+            "tags": self.tags(os_sg)
+        }
+
+        cidrs = self.get_compacted_cidrs(os_sg.get("cidrs"))
+        if cidrs:
+            sg["expression"].append({
+                "resource_type": "ConjunctionOperator",
+                "conjunction_operator": "OR"
+            })
+            sg["expression"].append({
+                "resource_type": "IPAddressExpression",
+                "ip_addresses": cidrs
+            })
+        return sg
+
+    def sg_rules_remote(self, cidr):
+        # NSX bug. Related IPSet to handle  0.0.0.0/x and ::0/x
+        return {
+            "display_name": cidr,
+            "expression": [{
+                "resource_type": "IPAddressExpression",
+                "ip_addresses": [cidr]   
+            }],
+            "tags": self.tags(None)
+        }
+
+    def sg_rules_container(self, os_sg, provider_sg):
+        return {
+            "category": "Application",
+            "display_name": os_sg.get("id"),
+            "stateful": True,
+            "tcp_strict": NSXV3_CAPABILITY_TCP_STRICT in os_sg.get("tags"),
+            "scope": [
+                "/infra/domains/default/groups/{}".format(
+                    provider_sg.get("scope"))
```

Comment:
I am a bit worried about the intense use of .get() on dictionaries:

For some critical values, like the scope, a missing key could hide an critical error and eventually cause issues by applying a security group to all ports. This is only hypothetical (I didn't check if thats the case in this example), but for absolutely required values as here I don't see why not just use `[key]`, which would make issues visible and abort the operation by throwing an exception.

Reply:
I am using it in order to create the full payload, then on error to know how it looked like.
In this case the request will fail due to mall formatted path.